packages/core/src/shared-priority-cache/shared-cache.ts

```diff
@@ -0,0 +1,143 @@
+import * as lo from 'lodash' // vite/astro went bonkers - not sure why this helps, core has always had lodash in it...
+import { PriorityCache, type Store, type Resource } from "./priority-cache";
+const { uniqueId } = lo
+// goal: we want clients of the cache to experience a type-safe interface -
+// they expect that the things coming out of the cache are the type they expect (what they put in it)
+// this is not strictly true, as the cache is shared, and other clients may use different types
+// also also - there will not be a 1:1 relation between items and
+type CacheInterface<Key, Value extends Record<string, Resource>> = {
+    get: (k: Key) => Value | undefined
+    has: (k: Key) => boolean
+    unsubscribeFromCache: () => void;
+    setPriorities: (low: Set<Key>, high: Set<Key>) => void;
+}
+type Fetcher = (sig: AbortSignal) => Promise<Resource>
+
+type ClientSpec<Key, Value extends Record<string, Resource>> = {
+    isValue: (v: Record<string, Resource | undefined>) => v is Value
+    cacheKeys: (item: Key) => { [k in keyof Value]: string }
+    onDataArrived?: (cacheKey: string) => void // todo very not helpful
+    fetch: (item: Key) => { [k in keyof Value]: (abort: AbortSignal) => Promise<Resource> }
+}
+type ObjectValue<T extends Record<string, any>> = T extends Record<string, infer Value> ? Value : never;
+
+type KV<T extends Record<string, any>> = readonly [keyof T, ObjectValue<T>]
+
+function entries<T extends Record<string, any>>(t: T): ReadonlyArray<KV<T>> {
+    return Object.entries(t) as ReadonlyArray<KV<T>>
+}
+function mapFields<R extends Record<string, any>, Result>(r: R, fn: (v: ObjectValue<R>) => Result): { [k in keyof R]: Result } {
+    return entries(r).reduce((acc, [k, v]) => ({ ...acc, [k]: fn(v) }), {} as { [k in keyof R]: Result })
+}
+
+type Client = {
+    low: Set<string>
+    high: Set<string>
+    notify: undefined | ((cacheKey: string) => void)
+}
+export class FancySharedCache {
```

Comment:
here is the shared cache - I should rename it... maybe just SharedPriorityCache?